latte/models/mine.py

```diff
@@ -0,0 +1,200 @@
+"""
+Implementation of the modular MINE model for estimating the mutual information
+between two arbitrary random variables.
+Based on:
+Belghazi, M. I., Baratin, A., Rajeswar, S., Ozair, S., Bengio, Y., Courville, A., & Hjelm, R. D. (2018).
+Mutual information neural estimation. 35th International Conference on Machine Learning, ICML 2018, 2, 864–873.
+
+The code is inspired by and based on https://github.com/gtegner/mine-pytorch.
+"""
+from enum import Enum
+from typing import Any, Union, Dict, Tuple, Optional
+
+import torch
+import torch.nn as nn
+import pytorch_lightning as pl
+from torch import Tensor
+
+
+EPSILON = 1e-6
+
+
+class MINEObjectiveType(Enum):
+    MINE = 1
+    F_DIV = 2
+    MINE_BIASED = 3
+
+
+class MINEBiasCorrection(torch.autograd.Function):
+    """
+    The custom running-mean-based correction factor to account for the bias of the
+    minibatch-based gradient estimate of the MINE loss function.
+    """
+
+    @staticmethod
+    def forward(ctx: Any, t_marginal: torch.Tensor, running_mean: torch.Tensor) -> Any:
+        ctx.save_for_backward(t_marginal, running_mean)
+        y = torch.logsumexp(t_marginal, 0) - torch.log(torch.tensor(t_marginal.shape[0]))
+        return y
+
+    @staticmethod
+    def backward(ctx: Any, grad_output: torch.Tensor) -> Tuple[Union[float, Any], None]:
+        t_marginal, running_mean = ctx.saved_tensors
+        corrected_grad = grad_output * t_marginal.exp().detach() / (running_mean + EPSILON) / t_marginal.shape[0]
+        return corrected_grad, None
+
+
+def mine_objective(
+    t: torch.Tensor, t_marginal: torch.Tensor, running_mean: torch.Tensor, alpha: float
+) -> Union[torch.Tensor, torch.Tensor]:
+    """The partially bias-corrected loss function which keeps a running mean
+    of the denominator of the second term from expression (12) in the paper."""
+
+    # detached such that the running mean computation does not affect the gradient, which are corrected manually
+    exp_t_marginal = torch.exp(torch.logsumexp(t_marginal, 0) - torch.log(torch.tensor(t_marginal.shape[0]))).detach()
+
+    running_mean = exp_t_marginal if running_mean == 0 else (1 - alpha) * running_mean + alpha * exp_t_marginal
+
+    y = MINEBiasCorrection.apply(t_marginal, running_mean)
+
+    return torch.mean(t) - y, running_mean
+
+
+def f_div_objective(t: torch.Tensor, t_marginal: torch.Tensor) -> torch.Tensor:
+    """The f-divergence upper-bound-based loss."""
+    return torch.mean(t) - torch.mean(torch.exp(t_marginal - 1))
```

Comment:
How numerically stable is this? Shall we try to move `t_marginal.max()` before the mean and subtract it from `t_marginal`?
(I'm unsure how the original implementation looks like, perhaps this works very well, so just asking :slightly_smiling_face: )

Reply:
Hmm, so the implementation here https://github.com/gtegner/mine-pytorch/blob/master/mine/models/mine.py is the same as I have. It's not the original authors but it's a very faithful implementation.
I don't think it's very important since I don't think we'll be using this method much any way, but did you mean replacing the second term with `torch.mean(torch.exp(t_marginal - torch.max(t_marginal) - 1)) * torch.max(t_marginal)`?

Reply:
Yes, this is what I meant. On the second thought, I'm not sure if it'd be more stable. As you know that it works, it's looks good for me as well – please, merge. Good job! :slightly_smiling_face: